### PR TITLE
Add hardCTF blogue

### PR DESCRIPTION
**Pull request by ruscur**

### CODE REVIEW

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+A few targets came to mind.  We could spray around where we think creds are
+located to try and overwrite the effective user ID of a process to 0, making it
+run as root.  We could go after something like SELinux, aiming for flags like
+`selinux_enabled` and `selinux_enforcing`.  I'm sure there's other sandbox-type
+controls we could try and escape from, too.
+
+None of these were taking my CTF in the direction I wanted it to go (which was
+shellcode running in the kernel), so I decided to turn the realism down a notch
+and aim for exploiting a null pointer dereference.  We'd map our shellcode to
+`*0`, induce a null pointer derefence in the kernel, and then our exploit would
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: dereference

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+vm.mmap_min_addr=0` and moved on for brevity.  As I was writing this I decided
+to make sure it was possible to bypass this without cheating by making use of
+our kernel write primitive, and sure enough, it works!  I had to zero out both
+`mmap_min_addr` and `dac_mmap_min_addr` symbols, the latter seemingly required
+for filesystem interactions to work post-exploit.
+
+So now we can trigger a null pointer dereference in the kernel and we can
+`mmap()` our shellcode to 0x0, we should probably get some shellcode.  We want
+to escalate our privileges, and the easiest way to do that is the iconic
+`commit_creds(prepare_kernel_cred(0))`.  As of somewhat recently it's a bit more
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: elaborate

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+After the CTF I encouraged everyone to try writing their own shellcode and noone
+did, and I will take that as a sign that mine is flawlessly designed.
+
+First we throw 0 into `r30`, which sounds like a register we'll get away with
+clobbering.  We load an offset of 32 bytes from the value of `r30` into `r12`
+(and `r30` is 0, so this is the address 32).  Then, we store the value of `r30`
+(0) into the address in `r12`.
+
+(we write zero to the address placed at 32 / 0x20)
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: this looks odd?

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+Then, we replace the contents of `r12` with the value contained at address 24.
+Then, we move that value into the count register, and branch to the count
+register.
+
+(indirect branch to the address placed at 24 / 0x18)
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: as does this

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+Overwriting have_canfork_callback...
+Successfully acquired root shell!
+/ # whoami
+root
+```
+
+It's wild to think that even an exploit this simple would have been possible in
+the "real world" back in 2017.  Still, it made for a good introduction to kernel
+exploitation for me and my team and wasn't *too* contrived for the sake of
+simplicity.
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: can we emphasise that this is why mitigations are valuable

**Review by ajdlinux — CHANGES_REQUESTED** [review]

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown (hunk 2 above)
Comment (ruscur) [reply content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: lazy but ok

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown (hunk 3 above)
Comment (ruscur) [reply content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: it does but I felt like the explanation was overcomplicated so I wanted to summarise it, I can try and do that in a different way but I'm not exactly sure how

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown (hunk 5 above)
Comment (ruscur) [reply content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: yeah ok

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+First we throw 0 into `r30`, which sounds like a register we'll get away with
+clobbering.  We load an offset of 32 bytes from the value of `r30` into `r12`
+(and `r30` is 0, so this is the address 32).  Then, we store the value of `r30`
+(which is 0) into the address in `r12` - writing zero to the address found at
+`*32`.
+
+Then, we replace the contents of `r12` with the value contained at address 24.
+Then, we move that value into the count register, and branch to the count
+register - redirecting execution the address found at `*24`.
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: to the address

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown [hunk 7]
```diff
@@ -0,0 +1,386 @@
+Title: Beginner's first kernel CTF with CVE-2017-5123
+Date: 2023-08-07 15:30:00
```
Comment (ajdlinux) [inline content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: date and filename now out of date

**Review by ajdlinux — CHANGES_REQUESTED** [review]

File: content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown (hunk 7 above)
Comment (ruscur) [reply content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown]: ree

**Comment by ruscur** [comment]
Having dates in filenames is extremely jank

**Pull request merged**